### ISSUE TITLE
请问post_prob里头self.cood是什么意思？能解释一下这部分的源代码吗？

### ISSUE THREAD

**Issue by guodewen** [ZhihengCV/Bayesian-Crowd-Counting#8]
请问您能解答一下【方括号标注部分】的计算原理么？感谢。
import torch
from torch.nn import Module

class Post_Prob(Module):
    def __init__(self, sigma, c_size, stride, background_ratio, use_background, device):
        super(Post_Prob, self).__init__()
        assert c_size % stride == 0

        self.sigma = sigma
        self.bg_ratio = background_ratio
        self.device = device
        # coordinate is same to image space, set to constant since crop size is same
        self.cood = torch.arange(0, c_size, step=stride,
                                 dtype=torch.float32, device=device) + stride / 2
【self.cood是用来做什么的】
        self.cood.unsqueeze_(0)
        self.softmax = torch.nn.Softmax(dim=0)
        self.use_bg = use_background

    def forward(self, points, st_sizes):
        num_points_per_image = [len(points_per_image) for points_per_image in points]
        # 每张图的人头数
        all_points = torch.cat(points, dim=0)
        # 人头坐标(n*2)
        # print(all_points.shape)
        if len(all_points) > 0:
            x = all_points[:, 0].unsqueeze_(1)
            # print(x.shape)
            # 横坐标列表
            y = all_points[:, 1].unsqueeze_(1)
            # print(y.shape)
            # 纵坐标列表
【这四行是计算什么？】
            x_dis = -2 * torch.matmul(x, self.cood) + x * x + self.cood * self.cood
            y_dis = -2 * torch.matmul(y, self.cood) + y * y + self.cood * self.cood
            y_dis.unsqueeze_(2)
            x_dis.unsqueeze_(1)
            dis = y_dis + x_dis
            dis = dis.view((dis.size(0), -1))

            dis_list = torch.split(dis, num_points_per_image)
            prob_list = []
            for dis, st_size in zip(dis_list, st_sizes):
                if len(dis) > 0:
                    if self.use_bg:
                        min_dis = torch.clamp(torch.min(dis, dim=0, keepdim=True)[0], min=0.0)
                        d = st_size * self.bg_ratio
                        bg_dis = (d - torch.sqrt(min_dis))**2
                        dis = torch.cat([dis, bg_dis], 0)  # concatenate background distance to the last
                    dis = -dis / (2.0 * self.sigma ** 2)
                    prob = self.softmax(dis)
                else:
                    prob = None
                prob_list.append(prob)
        else:
            prob_list = []
            for _ in range(len(points)):
                prob_list.append(None)
        return prob_list

**Comment by yuto3o** [comment]
关于loss我也有一些疑问
```
 def forward(self, prob_list, pre_density):
        loss = 0
        for idx, prob in enumerate(prob_list):  # iterative through each sample
            if prob is None:  # image contains no annotation points
                pre_count = torch.sum(pre_density[idx])
                target = torch.zeros((1,), dtype=torch.float32, device=self.device)
            else:
                N = len(prob)
                if self.use_bg:
                    target = torch.ones((N,), dtype=torch.float32, device=self.device)
                    target[-1] = 0.0  # the expectation count of background should be zero
                else:
                    target = torch.ones((N,), dtype=torch.float32, device=self.device)
                pre_count = torch.sum(pre_density[idx].view((1, -1)) * prob, dim=1)  # flatten into vector
        loss += torch.sum(torch.abs(target - pre_count))
        loss = loss / len(prob_list)
        return loss
```
 loss += torch.sum(torch.abs(target - pre_count)) 这行的缩进是不是错了

**Comment by ZhihengCV** [comment]
cood is the coordinate of each pixel

**Comment by ZhihengCV** [comment]
> 关于loss我也有一些疑问
> 
> ```
>  def forward(self, prob_list, pre_density):
>         loss = 0
>         for idx, prob in enumerate(prob_list):  # iterative through each sample
>             if prob is None:  # image contains no annotation points
>                 pre_count = torch.sum(pre_density[idx])
>                 target = torch.zeros((1,), dtype=torch.float32, device=self.device)
>             else:
>                 N = len(prob)
>                 if self.use_bg:
>                     target = torch.ones((N,), dtype=torch.float32, device=self.device)
>                     target[-1] = 0.0  # the expectation count of background should be zero
>                 else:
>                     target = torch.ones((N,), dtype=torch.float32, device=self.device)
>                 pre_count = torch.sum(pre_density[idx].view((1, -1)) * prob, dim=1)  # flatten into vector
>         loss += torch.sum(torch.abs(target - pre_count))
>         loss = loss / len(prob_list)
>         return loss
> ```
> 
> loss += torch.sum(torch.abs(target - pre_count)) 这行的缩进是不是错了

the loss is right, you can print the shape

**Comment by ZhihengCV** [comment]
> 关于loss我也有一些疑问
> 
> ```
>  def forward(self, prob_list, pre_density):
>         loss = 0
>         for idx, prob in enumerate(prob_list):  # iterative through each sample
>             if prob is None:  # image contains no annotation points
>                 pre_count = torch.sum(pre_density[idx])
>                 target = torch.zeros((1,), dtype=torch.float32, device=self.device)
>             else:
>                 N = len(prob)
>                 if self.use_bg:
>                     target = torch.ones((N,), dtype=torch.float32, device=self.device)
>                     target[-1] = 0.0  # the expectation count of background should be zero
>                 else:
>                     target = torch.ones((N,), dtype=torch.float32, device=self.device)
>                 pre_count = torch.sum(pre_density[idx].view((1, -1)) * prob, dim=1)  # flatten into vector
>         loss += torch.sum(torch.abs(target - pre_count))
>         loss = loss / len(prob_list)
>         return loss
> ```
> 
> loss += torch.sum(torch.abs(target - pre_count)) 这行的缩进是不是错了

thanks, the bug is fixed, this bug has no influence when batch_size=1
